MapboxNavigation/NavigationMapView.swift

```diff
@@ -671,6 +676,188 @@ open class NavigationMapView: MGLMapView, UIGestureRecognizerDelegate {
+                    if !routeShape.coordinates.isEmpty {
+                        return routeShape.coordinates
+                    } else { return [] }
+                } else {
+                    return []
+                }
+            }
+        }
+        let flatList = nestedList.flatMap { $0.flatMap { $0.flatMap { $0 } } }
+        return RoutePoints(nestedList: nestedList, flatList: flatList)
+    }
+    
+    /**
+    Find and cache the index of the upcoming [RouteLineDistancesIndex].
+    */
+    public func updateUpcomingRoutePointIndex(routeProgress: RouteProgress) {
+        let currentLegProgress = routeProgress.currentLegProgress
+        let currentStepProgress = routeProgress.currentLegProgress.currentStepProgress
+        if currentLegProgress != nil && currentStepProgress != nil && primaryRoutePoints != nil {
+            let completeRoutePoints = primaryRoutePoints!
+            /**
+            Find the count of remaining points in the current step.
+            */
+            var allRemainingPoints = getSlicedLinePointsCount(currentLegProgress: currentLegProgress, currentStepProgress: currentStepProgress)
+            
+            /**
+            Add to the count of remaining points all of the remaining points on the current leg, after the current step.
+            */
+            let currentLegSteps = completeRoutePoints.nestedList[routeProgress.legIndex]
+            if currentLegProgress.stepIndex < currentLegSteps.count {
+                let startIndex = currentLegProgress.stepIndex + 1
+                let endIndex = currentLegSteps.count - 1
+                allRemainingPoints += currentLegSteps.prefix(endIndex).suffix(from: startIndex).flatMap{ $0.flatMap{ $0 } }.count
+            }
+            
+            /**
+            Add to the count of remaining points all of the remaining legs.
+            */
+            for index in stride(from: routeProgress.legIndex + 1, to: completeRoutePoints.nestedList.count, by: 1) {
+                allRemainingPoints += completeRoutePoints.nestedList[index].flatMap{ $0 }.count
+            }
+            
+            /**
+            After calculating the number of remaining points and the number of all points,  calculate the index of the upcoming point.
+            */
+            let allPoints = completeRoutePoints.flatList.count
+            primaryRouteRemainingDistancesIndex = allPoints - allRemainingPoints - 1
+        } else {
+            primaryRouteRemainingDistancesIndex = nil
+        }
+    }
+    
+    private func getSlicedLinePointsCount(currentLegProgress: RouteLegProgress, currentStepProgress: RouteStepProgress) -> Int {
+        let startDistance = currentStepProgress.distanceTraveled
+        let stopDistance = currentStepProgress.step.distance
+        var slicedLine: Turf.LineString?
+        
+        /**
+        Implement the Turf.lineSliceAlong(lineString, startDistance, stopDistance) to return a sliced lineString.
+        */
```

Comment:
Please try to keep correct indentation for comments. To fix it you can select current comment and press <kbd>⌃</kbd> + <kbd>i</kbd>.